common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

```diff
@@ -404,6 +406,11 @@ public final void close() throws IOException {
         // Close the engine.
```

Comment:
What happens if `super.close();` throws on 403?

Reply:
Good point, I'll put `in.release()` in a finally block.

The `stateLock == null` check is unfortunately actually necessary, because the superclass constructor can call `close()` in some circumstances (see #404).  Feel free to complain to the authors of java.net.Socket about that.

Reply:
Ouch, that's no fun!